File: pwiz/analysis/spectrum_processing/SpectrumList_Smoother.cpp
//
// $Id$
//
//
// Original author: Matt Chambers <matt.chambers <a.t> vanderbilt.edu>
//
// Copyright 2008 Vanderbilt University - Nashville, TN 37232
//
// Licensed under the Apache License, Version 2.0 (the "License"); 
// you may not use this file except in compliance with the License. 
// You may obtain a copy of the License at 
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software 
// distributed under the License is distributed on an "AS IS" BASIS, 
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
// See the License for the specific language governing permissions and 
// limitations under the License.
//


#define PWIZ_SOURCE


#include "SpectrumList_Smoother.hpp"
#include "pwiz/analysis/common/SavitzkyGolaySmoother.hpp"
//#include "WhittakerSmoother.hpp"
#include "pwiz/utility/misc/Container.hpp"


namespace pwiz {
namespace analysis {


using namespace msdata;
using namespace pwiz::util;


PWIZ_API_DECL
SpectrumList_Smoother::SpectrumList_Smoother(
        const msdata::SpectrumListPtr& inner,
        SmootherPtr algorithm,
        const IntegerSet& msLevelsToSmooth)
:   SpectrumListWrapper(inner),
    algorithm_(algorithm),
    msLevelsToSmooth_(msLevelsToSmooth)
{
    // add processing methods to the copy of the inner SpectrumList's data processing
    ProcessingMethod method;
    method.order = dp_->processingMethods.size();
    method.set(MS_smoothing);
    method.userParams.push_back(UserParam("Savitzky-Golay smoothing (9 point window)"));
    
    if (!dp_->processingMethods.empty())
        method.softwarePtr = dp_->processingMethods[0].softwarePtr;

    dp_->processingMethods.push_back(method);
}


PWIZ_API_DECL bool SpectrumList_Smoother::accept(const msdata::SpectrumListPtr& inner)
{
    return true;
}


PWIZ_API_DECL SpectrumPtr SpectrumList_Smoother::spectrum(size_t index, bool getBinaryData) const
{
    //if (!getBinaryData)
    //    return inner_->spectrum(index, false);

    SpectrumPtr s = inner_->spectrum(index, true);

    vector<CVParam>& cvParams = s->cvParams;
    vector<CVParam>::iterator itr = std::find(cvParams.begin(), cvParams.end(), MS_profile_spectrum);

    // return non-profile spectra as-is
    if (itr == cvParams.end())
        return s;

    try
    {
        BinaryData<double>& mzs = s->getMZArray()->data;
        BinaryData<double>& intensities = s->getIntensityArray()->data;
        vector<double> smoothedMZs, smoothedIntensities;
        algorithm_->smooth(mzs, intensities, smoothedMZs, smoothedIntensities);
        mzs.swap(smoothedMZs);
        intensities.swap(smoothedIntensities);
        s->defaultArrayLength = mzs.size();
    }
    catch(std::exception& e)
    {
        throw std::runtime_error(std::string("[SpectrumList_Smoother] Error smoothing intensity data: ") + e.what());
    }

    s->dataProcessingPtr = dp_;
    return s;
}


} // namespace analysis 
} // namespace pwiz

File: pwiz-sharp/pwiz/src/Analysis/SpectrumList_Smoother.cs
using Pwiz.Analysis.PeakPicking;
using Pwiz.Data.Common.Cv;
using Pwiz.Data.Common.Params;
using Pwiz.Data.MsData.Processing;
using Pwiz.Data.MsData.Spectra;
using Pwiz.Util.Misc;

#pragma warning disable CA1707

namespace Pwiz.Analysis;

/// <summary>
/// SpectrumList wrapper that smooths profile-mode spectra via an <see cref="ISmoother"/>
/// (typically <see cref="SavitzkyGolaySmoother"/>). Centroid spectra pass through unchanged.
/// Port of <c>pwiz::analysis::SpectrumList_Smoother</c>.
/// </summary>
public sealed class SpectrumList_Smoother : SpectrumListWrapper
{
    private readonly ISmoother _algorithm;
    private readonly IntegerSet _msLevelsToSmooth;

    /// <summary>Constructs a smoothing wrapper around <paramref name="inner"/>.</summary>
    /// <param name="inner">The spectrum list to wrap.</param>
    /// <param name="algorithm">Smoothing algorithm.</param>
    /// <param name="msLevelsToSmooth">MS levels to apply smoothing to. Spectra at other levels
    /// pass through unchanged.</param>
    public SpectrumList_Smoother(ISpectrumList inner, ISmoother algorithm, IntegerSet msLevelsToSmooth)
        : base(inner)
    {
        ArgumentNullException.ThrowIfNull(algorithm);
        ArgumentNullException.ThrowIfNull(msLevelsToSmooth);
        _algorithm = algorithm;
        _msLevelsToSmooth = msLevelsToSmooth;
    }

    /// <inheritdoc/>
    public override DataProcessing? DataProcessing
    {
        get
        {
            var dp = new DataProcessing(Inner.DataProcessing?.Id ?? "pwiz_Reader_conversion");
            if (Inner.DataProcessing is not null)
                foreach (var pm in Inner.DataProcessing.ProcessingMethods)
                    dp.ProcessingMethods.Add(pm);
            var method = new ProcessingMethod
            {
                Order = dp.ProcessingMethods.Count,
                Software = dp.ProcessingMethods.FirstOrDefault()?.Software,
            };
            method.Set(CVID.MS_smoothing);
            method.UserParams.Add(new UserParam("Savitzky-Golay smoothing"));
            dp.ProcessingMethods.Add(method);
            return dp;
        }
    }

    /// <inheritdoc/>
    public override Spectrum GetSpectrum(int index, bool getBinaryData = false)
    {
        var spec = Inner.GetSpectrum(index, getBinaryData: true);

        // Skip non-profile spectra (centroided data is one peak per m/z; smoothing it would
        // average away the peaks themselves).
        if (!spec.Params.HasCVParam(CVID.MS_profile_spectrum))
            return spec;

        int msLevel = spec.Params.CvParamValueOrDefault(CVID.MS_ms_level, 0);
        if (!_msLevelsToSmooth.Contains(msLevel))
            return spec;

        var mzArr = spec.GetMZArray();
        var intArr = spec.GetIntensityArray();
        if (mzArr is null || intArr is null) return spec;

        var smoothedMz = new List<double>(mzArr.Data.Count);
        var smoothedInt = new List<double>(intArr.Data.Count);
        try
        {
            _algorithm.Smooth(mzArr.Data, intArr.Data, smoothedMz, smoothedInt);
        }
        catch (Exception e)
        {
            throw new InvalidOperationException(
                "[SpectrumList_Smoother] Error smoothing intensity data: " + e.Message, e);
        }

        // The two arrays should be the same length post-smooth — Smooth() pads xSmoothed via
        // ZeroSampleFiller to match the y output. Mirror cpp's swap-then-update-defaultArrayLength.
        spec.SetMZIntensityArrays(smoothedMz, smoothedInt, intArr.Params.CvParam(CVID.MS_intensity_unit).Cvid == CVID.CVID_Unknown
            ? CVID.MS_number_of_detector_counts
            : intArr.Params.CvParam(CVID.MS_intensity_unit).Cvid);
        spec.DefaultArrayLength = smoothedMz.Count;

        spec.DataProcessing = DataProcessing;
        return spec;
    }
}
